### DISCUSSION TITLE
Is there a memory leak in the hello world example?

### DISCUSSION THREAD

**Discussion by falkoschindler** [elimintz/justpy#552] [Q&A]
I modified the [Hello World example](https://justpy.io/#hello-world) a tiny bit to track the memory usage.

```python
#!/usr/bin/env python3
import gc
import os
import justpy as jp
import psutil

def hello_world():
    wp = jp.WebPage()
    process = psutil.Process(os.getpid())
    gc.collect()
    wp.add(jp.Div(text=f'memory: {process.memory_info().rss:,} bytes'))
    return wp

jp.justpy(hello_world)
```

It looks like there is a leak since the number keeps growing when reloading the page. This already happens with JustPy 0.2.x. Explicitly setting `temp=True` or `delete_flag=True` doesn't help.

Weirdly, `jp.WebPage.instances` are not growing. So, what else could hold the page or div objects? Or am I doing something completely wrong measuring the memory?

**Comment by WolfgangFahl** [comment]
@falkoschindler  We intend to have this in the #483 as well as the #541. So i appreciate i  a lot that you started the analysis. Would you have time for an online session so that we can add this feature to the demo browser together?

**Comment by WolfgangFahl** [comment]
see also https://stackoverflow.com/questions/552744/how-do-i-profile-memory-usage-in-python

**Comment by elimintz** [comment]
I ran your example on in my environment (Windows) and don't see a problem. The memory stabilizes or goes down after 20 or 30 reloads. The memory may be going up because python is allocating more memory to the dictionaries initially even though the entries into them are not growing. https://lerner.co.il/2019/05/12/python-dicts-and-memory-usage/

By the way, there is MEMORY_DEBUG option that you can set to True in justpy.env that gives you the following printout whenever a websocket disconnects:

```
************************
Elements:  0 {}
WebPages:  1 {22: WebPage(page_id: 22, number of components: 1, reload interval: None)}
Sockets:  0 {}
Memory used: 101,470,208
************************
************************
Elements:  0 {}
WebPages:  1 {23: WebPage(page_id: 23, number of components: 1, reload interval: None)}
Sockets:  0 {}
Memory used: 101,470,208
************************
************************
Elements:  0 {}
WebPages:  1 {24: WebPage(page_id: 24, number of components: 1, reload interval: None)}
Sockets:  0 {}
Memory used: 101,470,208
************************
************************
Elements:  0 {}
WebPages:  1 {25: WebPage(page_id: 25, number of components: 1, reload interval: None)}
Sockets:  0 {}
Memory used: 101,470,208
************************

``` 

It is in the code here: https://github.com/justpy-org/justpy/blob/master/justpy/justpy.py#L177

We may need more elaborate ways to debug memory since it is a crucial issue for scaling.

**Comment by falkoschindler** [comment]
@elimintz @WolfgangFahl I can't confirm that the memory stabilizes or goes down. After 300 reloads the number is still growing (at least on my Mac). Sure, Python doesn't allocate new memory for each new dictionary entry, but in chunks. So for a while it might look as if it's not growing anymore. But if you keep reloading, it eventually will.

```
************************
Elements:  0 {}
WebPages:  1 {315: WebPage(page_id: 315, number of components: 1, reload interval: None)}
Sockets:  0 {}
Memory used: 111,132,672
************************
************************
Elements:  0 {}
WebPages:  1 {316: WebPage(page_id: 316, number of components: 1, reload interval: None)}
Sockets:  0 {}
Memory used: 111,165,440
************************
```

So even the built-in memory analysis indicates some leakage.

**Comment by elimintz** [comment]
Ok, will check this more. This is interesting: https://github.com/tiangolo/fastapi/issues/4649
Could be a starlette issue. 
The dictionaries with the webpages, websockets and components are not growing. That is clear. And when you open multiple browser tabs and then close them you see that the dict entries decrease.

**Comment by elimintz** [comment]
I upgraded to the latest version of starlette. The problem does not completely go away but the memory leak is more rare. I am using python 3.10

**Comment by falkoschindler** [comment]
@elimintz I can confirm that. Well, if it's a starlette issue, I think we can't do much here.

**Comment by WolfgangFahl** [comment]
I moved it to 0.12 since i consider it important but not urgent and supporting the need for #483.
Justpy is IMHO  far from being a stable environment and we might rather go for frequent "reboots" of applications and recommend this before assuming we'd be able to get the production stability and security that some might expect.

**Comment by elimintz** [comment]
I've kept the program running for almost 25 hours now, and it looks like eventually starlette reclaims the memory. Memory usage went down from startup. I think there is no issue here to take care of.

**Comment by WolfgangFahl** [comment]
see 
- https://github.com/tiangolo/fastapi/issues/4649
- https://stackoverflow.com/questions/69225455/memory-leak-problem-related-with-fastapi-gunicorn-and-multiprocessing-pool
